File: 2017/Case.UngroupAll/Case.UngroupAll/UI/form_Main.cs
using Autodesk.Revit.DB;
using Case.UngroupAll.Data;
using Microsoft.VisualBasic.ApplicationServices;
using System;
using System.Diagnostics;
using System.Reflection;
using Autodesk.Revit.UI;

namespace Case.UngroupAll.UI
{
  public partial class form_Main : System.Windows.Forms.Form
  {

    private clsSettings _s;

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="s"></param>
    public form_Main(clsSettings s)
    {
      InitializeComponent();
      progressBar1.Hide();
      _s = s;
    }

    #region Form Controls & Events

    /// <summary>
    /// Startup
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void form_Main_Load(object sender, EventArgs e)
    {
      Text = string.Format(
        "Ungroup All v{0}",
        Assembly.GetExecutingAssembly().GetName().Version);

      checkBoxGroupsDetail.Checked = false;
      checkBoxGroupsDetail.Enabled = false;
      checkBoxGroupsModel.Checked = false;
      checkBoxGroupsModel.Enabled = false;

      if (_s.ModelGroups.Count > 0)
      {
        checkBoxGroupsModel.Checked = true;
        checkBoxGroupsModel.Enabled = true;
        checkBoxGroupsModel.Text = string.Format("{0} Model Groups", _s.ModelGroups.Count);
      }
      if (_s.DetailGroups.Count > 0)
      {
        checkBoxGroupsDetail.Checked = true;
        checkBoxGroupsDetail.Enabled = true;
        checkBoxGroupsDetail.Text = string.Format("{0} Detail Groups", _s.DetailGroups.Count);
      }

    }

    /// <summary>
    /// Logo
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void pictureBox1_Click(object sender, EventArgs e)
    {
      Process.Start("http://apps.case-inc.com/");
    }

    /// <summary>
    /// OK
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void buttonOk_Click(object sender, EventArgs e)
    {

      try
      {
        // Usage Tracking    
        AssemblyInfo m_a = new AssemblyInfo(Assembly.GetExecutingAssembly());
      }
      catch { }

      int m_total = 0;
      if (checkBoxGroupsDetail.Checked)
        m_total += _s.DetailGroups.Count;
      if (checkBoxGroupsModel.Checked)
        m_total += _s.ModelGroups.Count;

      // Progress Bar
      buttonOk.Hide();
      buttonCancel.Hide();
      progressBar1.Show();
      progressBar1.Minimum = 0;
      progressBar1.Maximum = m_total;
      progressBar1.Value = 0;

      int m_sicM = 0;
      int m_sicD = 0;

      using (Transaction t = new Transaction(_s.ActiveDoc))
      {
        if (t.Start("Ungrouping All Groups") == TransactionStatus.Started)
        {
          if (checkBoxGroupsDetail.Checked)
          {
            foreach (var x in _s.DetailGroups)
            {
              try
              {
                x.UngroupMembers();
                progressBar1.Increment(1);
                m_sicD++;
              }
              catch {}
            }
          }
          if (checkBoxGroupsModel.Checked)
          {
            foreach (var x in _s.ModelGroups)
            {
              try
              {
                x.UngroupMembers();
                progressBar1.Increment(1);
                m_sicM++;
              }
              catch { }
            }
          }
          t.Commit();

        }
      }

      if (m_sicM + m_sicD < 1)
      {
        using (TaskDialog td = new TaskDialog("Failed"))
        {
          td.TitleAutoPrefix = false;
          td.MainInstruction = "Ungrouping Groups Failed";
          td.Show();
        }
      }
      else
      {
        string m_msg = string.Format("{0} of {1} Model Groups Ungrouped\n{2} of {3} Detail Groups Ungrouped",
            m_sicM, _s.ModelGroups.Count, m_sicD, _s.DetailGroups.Count);

        using (TaskDialog td = new TaskDialog("Results"))
        {
          td.TitleAutoPrefix = false;
          td.MainInstruction = "Ungrouping Results:";
          td.MainContent = m_msg;
          td.Show();
        }
      }

      Close();

    }

    /// <summary>
    /// Close
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void buttonCancel_Click(object sender, EventArgs e)
    {
      Close();
    }

    private void checkBoxGroupsDetail_CheckedChanged(object sender, EventArgs e)
    {
      buttonOk.Enabled = (checkBoxGroupsDetail.Checked || checkBoxGroupsModel.Checked);
    }

    private void checkBoxGroupsModel_CheckedChanged(object sender, EventArgs e)
    {
      buttonOk.Enabled = (checkBoxGroupsDetail.Checked || checkBoxGroupsModel.Checked);
    }

    #endregion

  }
}

File: 2018/Case.HiddenParameterToParameter/Case.HiddenParameterToParameter/UI/form_Main.vb
Imports Autodesk.Revit.DB
Imports Autodesk.Revit.UI
Imports System.Windows.Forms
Imports System.Diagnostics
Imports [Case].HiddenParameterToParameter.API
Imports [Case].HiddenParameterToParameter.Data

Public Class form_Main

  Private _s As clsSettings
  Private _events As Boolean = False

  ''' <summary>
  ''' Constructor
  ''' </summary>
  ''' <param name="s"></param>
  ''' <remarks></remarks>
  Public Sub New(s As clsSettings)
    InitializeComponent()

    ' Widen Scope
    _s = s

  End Sub

#Region "Form Controls & Events"

  ''' <summary>
  ''' Startup
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub form_Main_Shown(sender As Object, e As EventArgs) Handles Me.Shown

    ' Form
    Text = "Hidden Parameter to Parameter v" & _s.Version
    ProgressBar1.Hide()
    ComboBoxCategory.DataSource = _s.Cats
    ComboBoxCategory.DisplayMember = "DisplayName"
    ComboBoxCategory.SelectedIndex = 0

    ' Enable Events
    _events = True

  End Sub

  ''' <summary>
  ''' Help Documentation
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonHelp_Click(sender As System.Object, e As System.EventArgs) Handles ButtonHelp.Click
    Process.Start("http://apps.case-inc.com/content/free-migrate-parameter-parameter")
  End Sub

  ''' <summary>
  ''' Show Parameters for Selected Category
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ComboBoxCategory_SelectedIndexChanged(sender As System.Object,
                                                    e As System.EventArgs) Handles ComboBoxCategory.SelectedIndexChanged

    ' Ignore When Events Disabled
    If _events = False Then Exit Sub

    ' Process the Category Object
    Dim m_c As clsCategory = ComboBoxCategory.SelectedItem
    If m_c.CatParamsText.Count = 0 And m_c.CatParamsElement.Count = 0 Then
      m_c.GetInstances()

      ' Warn when no elements
      If m_c.CatInstances.Count < 1 Then
        MsgBox("There are no element instances for this category in your model", MsgBoxStyle.Information, "No Elements")
      End If

      ComboBoxParamSource.DataSource = m_c.CatParamsElement
      ComboBoxParamTarget.DataSource = m_c.CatParamsText
      Try
        ComboBoxParamSource.SelectedIndex = 0
      Catch
      End Try
      Try
        ComboBoxParamTarget.SelectedIndex = 0
      Catch
      End Try
    End If

    ' Allow Port
    If ComboBoxParamSource.Items.Count = 0 Or ComboBoxParamTarget.Items.Count = 0 Then
      ButtonOk.Enabled = False
    Else
      ButtonOk.Enabled = True
    End If

  End Sub

  ''' <summary>
  ''' Commit
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonOk_Click(sender As System.Object,
                             e As EventArgs) Handles ButtonOk.Click

    ' Results
    Dim iTry As Integer = 0
    Dim iFail As Integer = 0

    Try

      ' The Selected Category
      Dim m_c As clsCategory = ComboBoxCategory.SelectedItem
      If m_c Is Nothing Then
        MsgBox("Failed to Get Selected Category Element Data :(",
               MsgBoxStyle.Critical,
               "Error")
        Exit Sub
      End If

      ' Form Setup
      ButtonOk.Hide()
      ButtonCancel.Hide()
      ButtonHelp.Hide()
      iTry = m_c.CatInstances.Count

      ' Progress Bar
      With ProgressBar1
        .Show()
        .Minimum = 0
        .Value = 0
        .Maximum = m_c.CatInstances.Count + 1
      End With

      ' Process Each Item
      For Each x In m_c.CatInstances

        ' Progress
        Try
          ProgressBar1.Increment(1)
          Application.DoEvents()
        Catch
        End Try

        ' Start a New transaction
        Using t As New Transaction(_s.Doc, "Element Parameters")
          If t.Start = TransactionStatus.Started Then

            Try

              Dim m_sourceValue As String = ""
              If ComboBoxParamSource.SelectedItem.ToString().ToLower() = "elementid" Then
                m_sourceValue = x.Id.ToString()
              Else
                Dim m_sourceParam As New clsParameter(x.LookupParameter(ComboBoxParamSource.SelectedItem.ToString))
                If m_sourceParam.ParameterObject Is Nothing Then Continue For
                If RadioButtonText.Checked = True Then
                  m_sourceValue = m_sourceParam.ValueString
                Else
                  m_sourceValue = m_sourceParam.Value
                End If
              End If
              If String.IsNullOrEmpty(m_sourceValue) Then Continue For
              Dim m_targetParam As New clsParameter(x.LookupParameter(ComboBoxParamTarget.SelectedItem.ToString))
              If m_targetParam.ParameterObject IsNot Nothing Then
                m_targetParam.Value = m_sourceValue
              End If

              ' Success
              t.Commit()

            Catch ex As Exception

              ' Failure
              iFail += 1
              t.RollBack()

            End Try

          End If
        End Using

      Next

    Catch ex As Exception

    End Try

    ' Report Results
    Using td As New TaskDialog("Element Update Results")
      td.MainInstruction = "Elements Updated!"
      td.MainContent = String.Format("Attempted to Update {0} Elements" & vbCr & vbCr & "{1} Failures",
                                     iTry.ToString, iFail)
      td.Show()
    End Using

    ' Close
    Close()

  End Sub

  ''' <summary>
  ''' Close
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonCancel_Click(sender As System.Object,
                                 e As System.EventArgs) Handles ButtonCancel.Click

    ' Close
    Close()

  End Sub

  ''' <summary>
  ''' Launch Case Site
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub PictureBox1_Click(sender As System.Object,
                                e As System.EventArgs) Handles PictureBox1.Click
    Process.Start("http://www.case-inc.com")
  End Sub

#End Region

End Class